### PR TITLE
[tensor-shapes] add stubs for jax.numpy einsum & tensor contractions

### PR DESCRIPTION
**Pull request by jakevdp**
### Summary

Add tensor shape stubs for `jax.numpy` einsum & other tensor contractions

### Test Plan

Updated existing unit tests, and tests pass locally:
```
$ uv tool run --from ruff==0.14.3 ruff format --check --config ruff.toml .
328 files already formatted
$ uv tool run --from ruff==0.14.3 ruff check --select I --config ruff.toml .
All checks passed!
$ uv run python test.py --no-test --no-conformance --no-jsonschema
Running formatting...
Finished in 2.69 seconds.
Running linting...
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.91s
Finished in 1.15 seconds.
$ uv run python tensor-shapes/pyrefly-jax-stubs/run_pyrefly.py --release
+ cargo build -p pyrefly --release
    Finished `release` profile [optimized + debuginfo] target(s) in 0.61s
PASS stubs (9 files)
PASS arithmetic (1 files)
PASS contractions (1 files)
PASS creation (1 files)
PASS fft (1 files)
PASS lax (1 files)
PASS linalg (1 files)
PASS matmul (1 files)
PASS nn (1 files)
PASS reductions (1 files)
PASS reshape (1 files)
```

### AI disclosure

Change generated with a Gemini coding agent.

### CODE REVIEW

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi [hunk 1]
```diff
@@ -886,9 +904,167 @@ def logaddexp2[Shape: _Shape](
...
 ) -> Array[broadcast(Shape1, Shape2)]: ...
+def cross[Shape1: _Shape, Shape2: _Shape](
+    a: Array[Shape1],
+    b: Array[Shape2],
+    /,
+    axisa: int = -1,
+    axisb: int = -1,
+    axisc: int = -1,
+    axis: int | None = None,
+) -> Array[broadcast(Shape1, Shape2)]: ...
```
Comment (stroxler) [inline tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi]: How correct is this? An agent thinks this is an approximation, but when `axis` arguments remap dimensions or there's a 2D vs 3D cross broadcast might not be right.

If that's the case, we might want to make a new DSL function and if necessary put a gradual fallback in place (you can use type vars with an upper bound of `Flag[int]` to capture arguments and pass them to the DSL, if you haven't yet used that, and the DSL allows you to explicitly return gradual shapes and sizes)

If it turns out to be hard to do well, I think a TODO is also fine (and if we discover the DSL is missing expressive power, happy to try to work on extending it! It's normal that we hit limitations in the DSL as we write more stubs)

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi [hunk 2]
```diff
@@ -494,3 +494,182 @@ def tridiagonal_diag_minus_one_shape(a_shape: IntTuple) -> IntTuple:
...
+    if d0 == d1:
+        return dsl.concat(shape[2:], dsl.IntTuple((d0,)))
+    if dsl.is_concrete_int(d0) and dsl.is_concrete_int(d1):
+        if d0 < d1:
+            return dsl.concat(shape[2:], dsl.IntTuple((d0,)))
+        return dsl.concat(shape[2:], dsl.IntTuple((d1,)))
+    return dsl.concat(shape[2:], dsl.IntTuple((dsl.Int.gradual(),)))
+
+@type_shape_dsl_function
+def trace_shape(shape: IntTuple, offset: int, axis1: int, axis2: int) -> IntTuple:
```
Comment (stroxler) [inline tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi:607]: Same note as `diagonal_shape`, the extra args here probably impact the output shape but we don't seem to be using them.

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi [hunk 3]
```diff
@@ -494,3 +494,182 @@ def tridiagonal_diag_minus_one_shape(a_shape: IntTuple) -> IntTuple:
...
+    if dims < 0:
+        return dsl.Invalid("tensordot dims must be non-negative")
+    if dims > len(left) or dims > len(right):
+        return dsl.Invalid("tensordot dims exceeds input rank")
+    if any(left[len(left) - dims + i] != right[i] for i in range(dims)):
+        return dsl.Invalid("tensordot contracted dimensions must match")
+    return dsl.concat(left[: len(left) - dims], right[dims:])
+
+@type_shape_dsl_function
+def diagonal_shape(shape: IntTuple, offset: int, axis1: int, axis2: int) -> IntTuple:
```
Comment (stroxler) [inline tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi:540]: An agent thinks this is probably approximate: offset and axis are ignored here, but they play a role in the correct result.

Maybe we could fall back to gradual shapes if it's hard to get the right thing, or at least put a TODO in to come back to this?

**Review by stroxler — CHANGES_REQUESTED** [review]
Overall it looks good to me.

Back to you for a few cases where I think the stubs are lossy without being gradual, I think we should try to improve them and/or throw some TODOs in

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi (hunk 1 above)
Comment (jakevdp) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi]: Updated!

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi (hunk 3 above)
Comment (jakevdp) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi:540]: Updated!

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi (hunk 2 above)
Comment (jakevdp) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/_shapes.pyi:607]: Updated!

**Review by stroxler — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.
